### PR TITLE
Onboarding check list

### PR DESCRIPTION
**Pull request by ghost**
#### Summary
This PR introduces the UI component for the onboarding checklist, that would be shown to first admin users. It will be shown in all of our products, hence the render in root and fixed positioning. Tested with Playbooks and it renders as needed. 

#### Ticket Link
https://mattermost.atlassian.net/wiki/spaces/GSI/pages/2271182849/Onboarding+Task+List
https://community-daily.mattermost.com/boards/workspace/fpbb7orqpiyfxqgg6fcu91y5xo/b6ropkdiowj8eugbt8kifrew79h/v1uasiye9zjg1fjfnbhq7muih6w/cwrz8edk5dfg8pg36crfa198itr

#### Screenshots
Screenshot 2022-02-06 at 19 34 32
Screenshot 2022-02-06 at 19 35 21
Screenshot 2022-02-06 at 19 34 23

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Review by matthewbirtch — COMMENTED** [review]
This is great work @nevyangelova. I've done a pretty thorough review and here's a list of polish items - pretty detailed stuff, but should all be pretty easy fixes. Let me know if you have any questions about anything.

1. For the list container, add  `border: solid 1px rgba(var(--center-channel-color-rgb),.16);`
2. Can we adjust the duration of the open animation on the task list to `150ms` (instead of 300ms)
3. Can we also adjust the duration of the background fade animation to `150ms`?
4. It looks like the wrong theme color is being used for the background of the task list container. It should be `background-color: var(--center-channel-bg);`
5. Can we reduce the margin above the 'Welcome to Mattermost' headline to `margin-top: 16px;`
6. Can we round the corners (4px radius) of the video thumbnail and add the 'Watch overview' button?
7. For each task in the list, the arrow icon should be 56% center channel color
8. The background color for each list item should be `background-color: var(--center-channel-bg);`
9. Task list items should be 36px in height if they are single-line text. Right now it looks like they are 44px. I think if we adjust the top/bottom padding to be 8px instead of 12px that should do it.
10. For task items where the text wraps, the icon should align with the first line - like the screenshot below 
image
11. Can we add 12px of padding on the top of the 'No thanks, I'll figure it out myself' text to separate it more from the task items?
12. Can we also add a hover state for the 'No thanks' line? (just add an underline on hover)
13 Also, can we use the `link-color` theme color for the 'No thanks' link?
14. For the Floating button, can we adjust the following?
    - default state
        - change background color to `background: var(--center-channel-bg);`
        - add  `border: solid 1px rgba(var(--center-channel-color-rgb),.16);`
        - add `box-shadow: 0px 6px 14px rgba(0, 0, 0, .16);`
        - icon color should be `rgba(var(--center-channel-color-rgb), .56)`
     - hover state
        -  `border: solid 1px rgba(var(--center-channel-color-rgb), .24);`
        - `box-shadow: 0px 8px 24px rgba(0, 0, 0, .16);`
        - icon color should be `rgba(var(--center-channel-color-rgb), .72)`
15. Can we also adjust the position of the number badge that appears on top of the floating button? I think it needs to move down 2px and to the right 2px
16. The number badge is maybe too small now that I see it (sorry, my bad in the design). Can we make the following changes on the number badge:
    width: 20px;
    height: 16px;
    background: var(--sidebar-text-active-border);
    position: fixed;
    display: block;
    border-radius: 12px;
    color: var(--button-color);
    font-size: 11px;
    line-height: 16px;
    margin-top: -31px;
    margin-left: 14px;
17. Can we reposition the floating button so that it is 20px from the bottom and 20px from the left edge?
18. Do you have the animation created for a completed task when the list is collapsed to the floating button?
19. Do you have the view created for when all the tasks in the list are complete?
20. I see that the 'Getting Started' section is there in the bottom left of the sidebar. The intention is the remove this from the experience for users who will see this list (since it does a lot of the same things). I'm not sure if this is part of what you are working on though.

**Comment by nevyangelova** [comment]
Thanks a lot @matthewbirtch for the thorough feedback it really helped me understand the design a bit better. Apologies for not showing the completed screen i missed adding the logic for it. I addressed all of your points but I still have some work to do with adjusting the completed animation based on the trigger. If you like you can take another look. After adjusting the paddings per the points above, I noticed that the container's height is smaller compared to the design. Should we keep it this way? 

The links and buttons currently won't work with dismissing the checklist I need to clarify where to store the user preference, and align with Boards and playbooks on this. And I will also remove the "Getting started" section when that is handled. Again thank you so much I am really grateful for the thorough feedback.

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
@nevyangelova I had another look and have a few more minor tweaks. I may have more, but didn't have a chance to review as closely today. Thought I'd send these along for now:

1. The icon on the floating button isn't quite centered within the circle 
image

2. For the floating button again, for some reason it seems like the entire button is not clickable. One some parts of it? Here's a gif of me hovering over different parts of the button. Notice the cursor change.


3. It looks like we still need more space above the 'No thanks, I'll figure it out myself' text to separate it more from the task items. For some reason the 12px you added isn't pushing it down. I think because it's a `span` it's not applying properly.

4. I was able to get the `all tasks completed` view. It looks like the content needs to be centered more though. It's sitting too low



5. Also on the tasks completed view, can we remove some of the left and right padding from headline and the paragraph so that they wrap on 2 lines instead of 3?

**Comment by matthewbirtch** [comment]
> Test server creation failed. See the logs for more information.

Can't test this right now @nevyangelova ☝️

**Comment by nevyangelova** [comment]
@matthewbirtch i fixed the issue :) note: don't dismiss the checklist because it won't show up again :D

File: components/global_onboarding_list/onboarding_checklist.tsx [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    border: solid 1px rgba(var(--center-channel-color-rgb), 0.16);
+    background-color: var(--center-channel-bg);
+    max-width: 352px;
+    padding: 1rem 0;
+    transform: scale(0);
+    opacity: 0;
+    box-shadow: 0px 20px 32px rgba(0, 0, 0, 0.12);
+    transition: opacity 150ms ease-in-out 0ms, transform 150ms ease-in-out 0ms;
+    transform-origin: left bottom;
+    height: 518px;
```
Comment (AshishDhama) [inline components/global_onboarding_list/onboarding_checklist.tsx:34]: Can we keep the height dynamic as there will be cases where it’s possible we will have less items in the list our more, Like I added Playbooks item in the list scrollbar appeared

File: components/global_onboarding_list/onboarding_checklist.tsx [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+            font-weight: bold;
+            font-size: 11px;
+            line-height: 16px;
+            bottom: 47px;
+            left: 41px;
+        }
+    `;
+});
+
+const PlayButton = styled.button`
```
Comment (AshishDhama) [inline components/global_onboarding_list/onboarding_checklist.tsx:111]: May be we can add some static width to button,as it is causing issue atleast on my system

File: components/global_onboarding_list/onboarding_checklist.tsx [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+                                />
+                            </h1>
+                            <p>
+                                <FormattedMessage
+                                    id='onboarding_checklist.main_subtitle'
+                                    defaultMessage="Let's get up and running."
+                                />
+                            </p>
+                            <img
+                                src={checklistImg}
```
Comment (AshishDhama) [inline components/global_onboarding_list/onboarding_checklist.tsx:230]: I think we should need to add to wrap inside a div with some color as placeholder, As image will take sometime to load and that’s creating issue.

File: components/root/root.jsx [hunk 4]
```diff
@@ -447,6 +451,7 @@ export default class Root extends React.PureComponent {
                     <CompassThemeProvider theme={this.props.theme}>
                         <ModalController/>
                         <GlobalHeader/>
+                        {this.props.isUserFirstAdmin && !this.props.isMobile && this.props.dismissChecklist && <TaskList/>}
```
Comment (AshishDhama) [inline components/root/root.jsx:454]:
```suggestion
                        {this.props.isUserFirstAdmin && !this.props.isMobile && !this.props.dismissChecklist && <TaskList/>}
```
as this value will not be there in preferences table, Although I think I can take care of this in my Pr

File: components/global_header/global_header.tsx [hunk 5]
```diff
@@ -22,7 +22,7 @@ const GlobalHeaderContainer = styled.header`
     border-bottom: solid 1px rgba(var(--center-channel-color-rgb), 0.08);
     color: rgba(var(--global-header-text-rgb), 0.64);
     padding: 0 12px;
-    z-index: 100;
+    z-index: 3;
```
Comment (AshishDhama) [inline components/global_header/global_header.tsx:25]: this is creating issue with Product menu, have fixed it mine as Temporary solution.
I think we should increase the z-index of overlay and Task list

**Review by AshishDhama — COMMENTED** [review]
Apart from these Rest look Good to me, Nice work on the Animations.

**Comment by matthewbirtch** [comment]
I can't see the task list on this PR anymore. I assume someone may have dismissed it or completed it and now we can't get it back. What do I need to do to reset it?

**Comment by esethna** [comment]
@matthewbirtch you probably have to remove and re-add the test server label so a new spinwick is created where the sys admin hasn't completed the list

File: components/global_header/global_header.tsx (hunk 5 above)
Comment (calebroseland) [reply components/global_header/global_header.tsx:25]: +1

File: components/global_onboarding_list/onboarding_checklist.tsx [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+        />,
+        <FormattedMessage
+            key={5}
+            id='onboarding_checklist.task_download_apps'
+            defaultMessage='Download the Desktop and Mobile Apps'
+        />,
+        <FormattedMessage
+            key={6}
+            id='onboarding_checklist.task_system_console'
+            defaultMessage='Visit the System Console to configure your worksace'
```
Comment (calebroseland) [inline components/global_onboarding_list/onboarding_checklist.tsx:191]:
```suggestion
            defaultMessage='Visit the System Console to configure your workspace'
```

**Comment by AshishDhama** [comment]
> I can't see the task list on this PR anymore. I assume someone may have dismissed it or completed it and now we can't get it back. What do I need to do to reset it?

@matthewbirtch @nevyangelova happening because of this condition https://github.com/mattermost/mattermost-webapp/pull/9727#discussion_r803933717

File: components/root/root.jsx (hunk 4 above)
Comment (calebroseland) [inline components/root/root.jsx:454]: Do we need a feature flag check here?

**Review by calebroseland — CHANGES_REQUESTED** [review]
Really great overall! Nice animations and visuals.

Only main additional issue I'm seeing is the invisible/covering div, otherwise just a couple nits/typos.

File: components/global_header/global_header.tsx (hunk 5 above)
Comment (nevyangelova) [reply components/global_header/global_header.tsx:25]: I can optimise that, however we are causing fundamental issues by assigning a z-index to the global header. Potentially we shouldn't have to pass a z-index at all to inline elements especially if they aren't covered by other elements. We should be adjusting the z-indexing of absolute positioned elements only, pop ups, menus, overlays. I see a lot of issues when I was trying to implement menus, be it using popper or not, when global header has such a high z-index. 

As for what is covering the menu in Products when rendered in boards, does that banner need to have a higher z-index too? I did a small test removing it and it didn't change the UI. Can we try to remove redundant z-indexes so we can prevent issues with overlap in the future?

File: components/global_onboarding_list/onboarding_checklist.tsx (hunk 1 above)
Comment (nevyangelova) [reply components/global_onboarding_list/onboarding_checklist.tsx:34]: ill convert to max-height in order to keep the wrapper on the next page consistent.

File: components/global_onboarding_list/onboarding_checklist.tsx (hunk 2 above)
Comment (nevyangelova) [reply components/global_onboarding_list/onboarding_checklist.tsx:111]: added    max-width: 175px;

File: components/root/root.jsx (hunk 4 above)
Comment (nevyangelova) [reply components/root/root.jsx:454]: missed it, sorry bout that guys :)

**Review by Willyfrog — APPROVED** [review]

File: package.json [hunk 7]
```diff
@@ -63,6 +63,7 @@
     "react-intl": "5.20.10",
     "react-is": "16.13.1",
     "react-overlays": "1.2.0",
+    "react-popper": "2.2.5",
```
Comment (Willyfrog) [inline package.json:66]: @AshishDhama did we end up using popper?

Comment (AshishDhama) [reply package.json:66]: @Willyfrog Nevy is using it for task list, Didn’t had discussions with her on this.

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Another few things I just noticed:
- Can we add 8px of space between the video and the first task list item? It's too tight together there.
Screen Shot 2022-02-11 at 9 59 27 AM

- Also, the height of the container for the task list doesn't grow with the content. It seems to be a fixed height for some reason. We'll need to adjust this since some cases will not have all of the items populated in the list (e.g. if admin doesn't select project management or workflows from the onboarding wizard). I think because of the min-height that is applied (and likely not necessary).

**Review by calebroseland — APPROVED** [review]

File: i18n/en.json [hunk 8]
```diff
@@ -3832,6 +3833,18 @@
...
+  "onboarding_checklist.completed_title": "Well done. You’ve completed all of the tasks!",
+  "onboarding_checklist.dismiss_link": "No thanks, I’ll figure it out myself",
+  "onboarding_checklist.documentation_link": " See our documentation.",
+  "onboarding_checklist.main_subtitle": "Let's get up and running",
+  "onboarding_checklist.task_boards_tour": "Manage tasks with your first board",
+  "onboarding_checklist.task_channels_tour": "Take a tour of channels",
+  "onboarding_checklist.task_complete_profile": "Complete your profile",
+  "onboarding_checklist.task_download_apps": "Download the Desktop and Mobile Apps",
+  "onboarding_checklist.task_invite": "Invite team members to the workspace",
+  "onboarding_checklist.task_system_console": "Visit the System Console to configure your worksace",
```
Comment (calebroseland) [inline i18n/en.json:3846]:
```suggestion
  "onboarding_checklist.task_system_console": "Visit the System Console to configure your workspace",
```

**Review by AshishDhama — APPROVED** [review]

**Comment by AshishDhama** [comment]
@matthewbirtch Could you please approve this will be Merging this pr to unblock pr dependent on this? (have changed show condition so that it doesn’t interfere with anything on the master). Changes requested here will be done via separate pr.

**Comment by AshishDhama** [comment]
e2e-test

**Review by matthewbirtch — APPROVED** [review]
Approving to unblock the merge. Any further tweaks can happen in follow-up PRs

**Pull request merged**